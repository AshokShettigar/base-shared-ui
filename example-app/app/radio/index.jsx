import CheckBoxExample from 'app/checkbox';
import { Field, Radio } from 'platform-common-ui';

function resolveSelected(options, selectedOption) {
    return options.map(option => ({
        ...option,
        isChecked: option.value === selectedOption
    }));
}

export default class RadioExample extends CheckBoxExample {
    constructor(props) {
        super(props);

        this.onAddNewOption = () => {
            const { newOption } = this.state;
            const tmpOptions = this.state.options.concat(newOption);
            const selectedOption = newOption.isChecked
                ? newOption.value
                : this.state.selectedOption;
            const options = resolveSelected(tmpOptions, selectedOption);
            this.setState({
                options,
                selectedOption,
                newOption: {
                    label: '',
                    value: '',
                    hideLabel: false,
                    isChecked: false,
                    isDisabled: false
                }
            });
        };
        this.onOptionIsCheckedChange = () => selectedOption => {
            const options = resolveSelected(this.state.options, selectedOption);
            this.setState({ options, selectedOption });
        };
    }

    renderComponent() {
        return (
            <Field
                hasError={this.state.hasError}
                isDisabled={this.state.isDisabled}
                isRequired={this.state.isRequired}
                label="Field option example"
            >
                {
                    this.state.options.map((option, index) => (
                        <Radio
                            hideLabel={option.hideLabel}
                            id={option.value}
                            isChecked={this.state.selectedOption === option.value}
                            isDisabled={option.isDisabled}
                            key={option.value}
                            onChange={this.onOptionIsCheckedChange(index)}
                            value={option.value}
                        >
                            {option.label}
                        </Radio>
                    ))
                }
            </Field>
        );
    }
}
