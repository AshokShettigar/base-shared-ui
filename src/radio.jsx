import { CheckBox } from 'platform-common-ui/checkbox';

// eslint-disable-next-line import/prefer-default-export
export const Radio = props => <CheckBox {...props} type="radio" />;
