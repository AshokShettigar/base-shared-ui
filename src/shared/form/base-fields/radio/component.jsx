import React, { PureComponent } from 'react';
import { string, bool, func, node } from 'prop-types';
import classNames from 'classnames';
import uuid4 from 'uuid';

import { FieldWithError } from '../index';
import styles from './component.scss';
import { withRender } from '../../../shared';

class Radio extends PureComponent {
    static propTypes = {
        id: string,
        name: string,
        value: string,

        className: string,
        labelClassName: string,
        containerClassName: string,

        checked: bool,
        disabled: bool,
        hidden: bool,

        children: node.isRequired,
        error: string,

        innerContainerRef: func,
        onChange: func
    };

    static defaultProps = {
        id: undefined,
        name: undefined,
        value: undefined,

        className: undefined,
        containerClassName: undefined,

        checked: false,
        disabled: false,
        hidden: false,

        error: undefined,

        innerContainerRef: undefined,
        onChange: undefined
    };

    onChange = e => {
        e.preventDefault();

        if (this.props.onChange && !this.props.disabled) {
            this.props.onChange(this.props.name, this.props.value);
        }
    };

    render () {
        const className = classNames(styles.root, this.props.className, {
            [styles.disabled]: this.props.disabled,
            [styles.checked]: this.props.checked
        });
        const id = this.props.id || uuid4();

        return (
            <span className={className} ref={this.props.innerContainerRef}>
                <FieldWithError error={this.props.error}>
                    <label htmlFor={id} className={classNames(styles.label, this.props.labelClassName)}>
                        <input
                            hidden
                            id={id}
                            type="radio"
                            name={this.props.name}
                            value={this.props.value}
                            checked={this.props.checked}
                            disabled={this.props.disabled}
                            onChange={this.onChange}
                        />
                        <span
                            role="presentation"
                            className={styles.radio}
                            onClick={this.onChange}
                        />
                        {this.props.children &&
                            <span
                                tabIndex={-1}
                                role="radio"
                                aria-checked={this.props.checked}
                                className={classNames(styles.container, this.props.containerClassName)}
                                onClick={this.onChange}
                            >
                                {this.props.children}
                            </span>
                        }
                    </label>
                </FieldWithError>
            </span>
        );
    }
}

export default withRender()(Radio);
