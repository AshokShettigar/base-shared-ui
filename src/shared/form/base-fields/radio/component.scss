@import '../../../scss/variables';

.root {
  display: inline-flex;
  flex-direction: column;
  font-family: $ff-open-sans;
  font-size: $fs-tiny;
  color: $c-raisin-black;
}

.label {
  display: flex;
  align-items: center;
  cursor: pointer;
  margin: 0;
  font-weight: normal;
}

.radio {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  width: 12px;
  height: 12px;
  border: 1px solid $c-gainsboro;
  border-radius: 100%;
  background-color: $c-white;

  &::after {
    content: "";
    width: 4px;
    height: 4px;
    border-radius: 100%;
  }
}

.checked .radio {
  background-color: $c-french-blue;
  border: 1px solid transparent;

  &::after {
    background-color: $c-white;
  }
}

.disabled {
  .label {
    cursor: default;
  }

  .radio {
    border-color: $c-gainsboro;
    background-color: $c-white;
  }
}

.disabled.checked .radio {
  &::after {
    background-color: $c-gainsboro;
  }
}

.container {
  line-height: 12px;
  margin-left: 5px;
  color: $c-raisin-black;
  user-select: none;
  display: flex;
}
