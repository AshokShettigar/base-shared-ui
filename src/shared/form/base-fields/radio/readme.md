## Radio
#### Usage
```JSX
import { shared } from 'platform-common-ui';
const { Radio } = shared;
```
```JSX
<Radio id="some-id" />
```

#### Description
Radio component permits the user to make a single choice.

#### Properties
Name                  | Type               | Required | Default     | Description       
:---------------------| :----------:       | :------: | :-----:     | :-----------
id                    | `string`           | `-`      | uuid()      | Setup the Radio's id
name                  | `string`           | `-`      | undefined   | Setup the Radio's name
value                 | `string`           | `-`      | undefined   | Setup the Radio's value
children              | `node`             | `-`      | undefined   | Displays the Radio's content
error                 | `string`           | `-`      | undefined   | Displays the Radio's error
checked               | `bool`             | `-`      | false       | Displays the checked Radio 
disabled              | `bool`             | `-`      | false       | Displays the disabled Radio 
hidden                | `bool`             | `-`      | false       | Makes the Radio's hidden
rendered              | `bool`             | `-`      | true        | Render Button to the DOM
className             | `string`           | `-`      | undefined   | Allows customizing styling of the component
labelClassName        | `string`           | `-`      | undefined   | Allows customizing styling of the Radio's label
containerClassName    | `string`           | `-`      | undefined   | Allows customizing styling of the Radio's content
innerContainerRef     | `func`             | `-`      | undefined   | Get a reference to the DOM component
onChange              | `func`             | `-`      | undefined   | Runs when the value of the field changes
