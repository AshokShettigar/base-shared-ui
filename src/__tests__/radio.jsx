import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { Radio } from 'platform-common-ui/radio';

describe('radio', function() {
    const onChange = () => true;

    it('should render the radio', function() {
        const tree = renderer.create(
            <Radio id="rd1" onChange={onChange} value="rd1_value">Check me!</Radio>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the radio with classname', function() {
        const tree = renderer.create(
            <Radio
                className="my-class"
                id="rd1"
                onChange={onChange}
                value="rd1_value"
            >
                Check me!
            </Radio>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the radio with disabled', function() {
        const tree = renderer.create(
            <Radio id="rd1" isDisabled onChange={onChange} value="rd1_value">Check me!</Radio>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the checkbox with no label', function() {
        const tree = renderer.create(
            <Radio id="rd1" hideLabel onChange={onChange} value="rd1_value">Check me!</Radio>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should render the radio with checked', function() {
        const tree = renderer.create(
            <Radio id="rd1" isChecked onChange={onChange} value="rd1_value">Check me!</Radio>
        );
        expect(tree).toMatchSnapshot();
    });

    it('should trigger onChange handler', function(done) {
        const onChangeHandler = value => {
            expect(value).toBe('rd1_value');
            done();
        };

        const event = {
            target: { value: 'rd1_value' }
        };
        const wrapper = mount(
            <Radio id="rd1" onChange={onChangeHandler} value="rd1_value">Check me!</Radio>
        );
        wrapper.find('input').simulate('change', event);
    });

    it('should throw console error when there is no children', function() {
        // eslint-disable-next-line no-console
        console.error = jest.fn();
        renderer.create(<Radio id="rd1" onChange={onChange} value="rd1_value" />);
        // eslint-disable-next-line no-console
        expect(console.error).toHaveBeenCalled();
    });
});
