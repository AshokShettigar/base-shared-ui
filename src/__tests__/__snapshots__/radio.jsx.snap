// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`radio should render the checkbox with no label 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="rd1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="rd1"
    onChange={[Function]}
    type="radio"
    value="rd1_value"
  />
  <span
    className="cn-field__text a11y"
  >
    Check me!
  </span>
</label>
`;

exports[`radio should render the radio 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="rd1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="rd1"
    onChange={[Function]}
    type="radio"
    value="rd1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`radio should render the radio with checked 1`] = `
<label
  className="cn-field__opt cn-field__opt--is-checked"
  data-component="CheckBox"
  htmlFor="rd1"
>
  <input
    checked={true}
    className="cn-field__checkradio"
    disabled={false}
    id="rd1"
    onChange={[Function]}
    type="radio"
    value="rd1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`radio should render the radio with classname 1`] = `
<label
  className="cn-field__opt my-class"
  data-component="CheckBox"
  htmlFor="rd1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={false}
    id="rd1"
    onChange={[Function]}
    type="radio"
    value="rd1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;

exports[`radio should render the radio with disabled 1`] = `
<label
  className="cn-field__opt"
  data-component="CheckBox"
  htmlFor="rd1"
>
  <input
    checked={false}
    className="cn-field__checkradio"
    disabled={true}
    id="rd1"
    onChange={[Function]}
    type="radio"
    value="rd1_value"
  />
  <span
    className="cn-field__text"
  >
    Check me!
  </span>
</label>
`;
